Feature/dashboard

dsnider opened the pull request:
WIP for Dashboard #40

Waffle relationships: closes #49, closes #42, connects #40, connects #52, connects #46

dsnider committed:
Rename ex_admin_test to talon_test

dsnider committed:
Add dashboard test

dsnider committed:
Add mix test watch - closes #42

dsnider committed:
Add gettext to mix #46

dsnider committed:
Add compilers for test env

dsnider committed:
Fix assert_file false positive tests - closes #49

dsnider committed:
Use list approach for assert_file #49

dsnider committed:
Fix typo

dsnider committed:
Ignore editor files

dsnider committed:
Remove model case

dsnider committed:
Change context file name to talon.ex

dsnider committed:
Finish setting up test fixtures

dsnider committed:
Fixup dashboard test

dsnider committed:
Remove TestHelpers import

dsnider committed:
Add test for dashboard view generation

dsnider committed:
Use assert_file list approach

dsnider committed:
Add dashboard template test

dsnider committed:
Add dashboard tests to talk new test

dsnider committed:
Add gen templates for dashboard page, view, template and page controller

dsnider committed:
Add dashboard to Talon config

dsnider committed:
Add first pass at Page, PageController #52, #40

dsnider committed:
Make View page aware #52, #40

dsnider committed:
Add dashboard paths to Resource #40

dsnider committed:
Use regular View plug for page controller #52, #40

dsnider committed:
Add initial page and dashboard support to Talon context #52, #40

dsnider committed:
Move schema handling from talon resource to load resource plug #52, #40

dsnider committed:
Fix router for page/dashboard handling #52, #40

dsnider committed:
Make app layout and sidebar page aware #52, #40

dsnider committed:
Add Gettext module for now

dsnider committed:
Fix typo in Talon new test

dsnider committed:
Add some code comments

dsnider committed:
Add page controller and dashboard gen to talk new #52, #40

dsnider committed:
Add dashboard template and view to theme gen #40

dsnider committed:
Tweak logic #40

dsnider committed:
Tweak .gitignore

dsnider committed:
Remove dependency on schema from View plug #52

dsnider committed:
Default Talon dashboard to nil, add comment

dsnider committed:
Tweak Talon dashboard_name

dsnider committed:
Tweak implementation and remove cruft from PageController

dsnider committed:
Remove extra spaces in Controller and remove comment from Router

dsnider committed:
Add Dashboard fixture and use for dashboard test

dsnider committed:
Ensure either talon_resource or talon_page is present in TalonResource plug

dsnider committed:
Test for dashboard under /pages

dsnider committed:
Add Talon.new dashboard tests for older Phoenix generator #40

dsnider committed:
Remove cruft from Page #52

dsnider committed:
Fix typo in resource

dsnider committed:
Test for dashboard template in phoenix structure

dsnider committed:
Fix some compiler warnings and typos

dsnider committed:
Refactor lib_path def, remove from talon new

dsnider committed:
Tweak some code comments

dsnider committed:
Remove dashboard setting from test config for now #40

dsnider committed:
Temp fix for dashboard gen #40

smpallen99 commented on lib/mix/tasks/talon.gen.theme.ex:
```diff
@@ -220,6 +223,41 @@ defmodule Mix.Tasks.Talon.Gen.Theme do
 
+  def gen_dashboard_template(%{dashboard: true} = config) do
+    binding = Kernel.binding() ++
+      [title_txt: gettext("Dashboard"),
```
Using gettext here does not work since you are evaluating the macro during talon compile. Instead, pass a string containing gettext so the macro call is imbedded in the generated. For example: `title_txt: ~s[gettext("Dashboard")]`

dsnider replied:
Okay, I was wondering exactly how we wanted to do that. I was looking at what Jose was saying here: https://github.com/smpallen99/coherence/issues/185

dsnider replied:
Should this use `dgettext` or `Talon.Messages` so the domain is picked up?

smpallen99 replied:
this should be coded at `[title_txt: ~s/dgettext("Talon", "Dashboard")/]`. This way the generated code contains a call to dgettext("Talon", "Dashboard") which is compiled in the context of the project, and therefore uses the projects Gettext module.

dsnider replied:
Sounds good about using `dgettext` but in this case, it doesn't appear we're actually using `'title_txt` anywhere... It was carried over from ExAdmin, but I didn't see it used there either. 

We can use dgettext on `welcome_txt` and `add_txt`, but that text is just placeholder information for the Talon developer and would typically deleted, no?

smpallen99 commented on lib/mix/tasks/talon.gen.theme.ex:
```diff
@@ -72,6 +72,7 @@ defmodule Mix.Tasks.Talon.Gen.Theme do
   use Mix.Task
 
   import Mix.Talon
+  import Talon.Gettext
```
We don't want to have a Talon.Gettext. We need to be using the project's gettext. See notes below.

dsnider replied:
Definitely, I have an internal TODO regarding this.

smpallen99 commented on lib/talon.ex:
```diff
@@ -52,11 +52,15 @@ defmodule Talon do
-      @__resources__  Application.get_env(:talon, :resources, [])
+      @__resources__  Application.get_env(:talon, :resources, []) # TODO: make easier to inject/test (DJS)
+
+      @__dashboard__  Application.get_env(:talon, :dashboard, nil) # TODO: support more than one dashboard (DJS)
```
I think this should be `@__pages__ Appplication.get_env(:talon, :pages, [])`

dsnider replied:
I agree, see below - I had a TODO on `@__page_map__` for this (with "page" incorrectly singular). I still wasn't certain if we shouldn't maintain a list of pages that are dashboards, as they get specific treatment when presented (i.e., listed first, before resources). Other pages may occur after resources. 

Note, I was leaning towards the idea of supporting multiple dashboards, if desired. 

Take look here, for example: https://adminlte.io/themes/AdminLTE/index2.html

Instead of first-class handling of  dashboards, this could be handled with pages and a concept of page_groups, where `dashboards` are one such pre-defined group. In the layout, the user can then decide where to put a given page group in the sidebar menu (e.g., dashboard group, resources, other page groups)

We could have simple default heuristic for presenting page groups (see example AdminLTE above):
- One item, just display page name without a container (current way we display dashboard link)
- Less than N, present as titled container of child page links, but always expanded
- More than N, present as collapsable container of children

smpallen99 commented on lib/talon.ex:
```diff
@@ -77,6 +81,17 @@ defmodule Talon do
 
       def resource_names, do: @__resource_map__ |> Map.keys
 
+      def dashboard_names, do: [__MODULE__. dashboard_name]
```
`page_names` ??

dsnider replied:
Yeah, I was cheating out of the gate and was still deciding on how dashboards and pages would ultimately relate (see above), so I went with just `dashboard_names` until I had a chance to revisit.

smpallen99 commented on lib/talon/gettext.ex:
```diff
@@ -0,0 +1,6 @@
+defmodule Talon.Gettext do
```
Per previous comments. Remove this.

smpallen99 commented on lib/talon/page.ex:
```diff
@@ -0,0 +1,59 @@
+      Note: This function is overridable
+      """
+      @spec dashboard_paths(Map.t) :: [Tuple.t]
+      def dashboard_paths(%{talon: talon} = _talon) do
```
`pages_paths` ??

smpallen99 commented on lib/talon/resource.ex:
```diff
@@ -167,7 +176,7 @@ defmodule Talon.Resource do
         talon.resources()
         |> Enum.map(fn talon_resource ->
-          schema = talon_resource.schema()
+          schema = talon_resource.schema() # TODO: let resource determine presentation name (DJS)
```
I believe your comment is more applicable to the line below. The `Talon.Utils.titleize(schema)` should be changed to be an overridable function call.

dsnider replied:
Yes, that's actually meant for the line below but the line I used was much shorter, so I cheated :)

smpallen99 commented on priv/templates/talon.gen.theme/admin_lte/templates/layout/sidebar.html.slim:
```diff
@@ -1,11 +1,19 @@
+      / TODO: use talon_context to get page/dashboard_paths (DJS)
+      - resource_or_page = @talon_resource || @talon_page
+      = if resource_or_page do
+        = for dashboard_path <- resource_or_page.dashboard_paths(@conn.assigns.talon) do
```
We should have a UI discussion for this. I'm thinking the UI should be:

Pages
  Dashboard
  MyPage1
  ...
Resources
  Users
  Posts
  ...
Actions
  ...

And therefore, we should be calling a `pages_paths` function to return the collection of pages.

dsnider replied:
Ah, I see.  I was thinking a different order as well as semi-first class treatment for dashboards (see my comments above).  Also, that feels a little flat for more complex applications and not easily scalable to handle grouping. Perhaps my suggestion of page groups could be used here for flexibility? Resource groups could also make sense.

smpallen99 reviewed (commented):
Well done Dan! Not very often that I review someone's work without making "style" related comments. I have none on your commits. I'm impressed!!

I think we should have a conversation about gettext, especially if you don't "get" my comments.

After reviewing your comments, I'm of the opinion that `resources` and `pages` as the two talon managed types works. I still think that dashboard is just the default name for a default (optional) landing page. 

I see a few areas in the code where you use `resource_or_page` and I think that is fine. I'm not sure we need a new word for that. Its more explicit than a creating a collection word. 

I'm open for discussion on this point, but wanted to share my initial reaction.

I have not commented on any of the areas where changes are going to be necessary to resolve merge conflicts, but there are many.

dsnider commented:
Thanks, Steve!

> I think we should have a conversation about gettext, especially if you don't "get" my comments.

Yep, definitely on board regarding `gettext`, but we may want to chat about exact solution (see comments).

Regarding `resources` and `pages`, I'm still thinking it would be good to decide what they are, both from taxonomy standpoint and architecturally. That allows us to refer to them collectively in documentation, discussion, in code, etc. By deciding a name, we can better establish rules regarding the primary responsibility of those modules, and from that, maintain cohesion.

dsnider committed:
Tighten page support and implement dashboards more directly as pages #40, #52

dsnider committed:
Remove Gettext and use target project’s Gettext

dsnider committed:
Use dgettext and remove title_txt from dashboard template gen #40

dsnider committed:
Don’t pluralize page titles

dsnider committed:
Tighten page implementation #52

smpallen99 commented on lib/talon/page_controller.ex:
```diff
@@ -0,0 +1,26 @@
+      end
+
+      @spec dashboard(Plug.Conn.t, Map.t) :: Plug.Conn.t
+      def dashboard(conn, params) do
```
Not sure we want a dashboard function here. If dashboard is like any other page, then the functions should be generic.

smpallen99 commented on lib/talon/plug/talon_resource.ex:
```diff
@@ -8,18 +8,18 @@ defmodule Talon.Plug.TalonResource do
+
     talon_resource = context.talon_resource(conn.params["resource"])
-    assign conn, :talon, Enum.into([talon_resource: talon_resource, schema: schema], talon)
+    talon_page = if talon_resource in [nil, []], do: context.talon_page(conn.params["page"])
```
need to check if we can use `page` as params key here. May get confused with page from pagination. Not sure, but just raising the point.

smpallen99 commented on lib/talon/plug/talon_resource.ex:
```diff
@@ -8,18 +8,18 @@ defmodule Talon.Plug.TalonResource do
+
+    talon_resource || talon_page || raise("resource or page required")
+
+    assign conn, :talon, Enum.into([talon_resource: talon_resource, talon_page: talon_page], talon)
```
these resource and page are mutually exclusive. So I wonder if we want one key. Yes, I get the naming dilemma here.

smpallen99 commented on lib/talon/resource.ex:
```diff
@@ -159,6 +159,17 @@ defmodule Talon.Resource do
+      Note: This function is overridable
+      """
+      @spec page_paths(Map.t) :: [Tuple.t]
+      def page_paths(%{talon: talon} = _talon) do
```
I can see the naming problem here. I'm starting to warm up to a collective name.

smpallen99 commented on lib/talon/router.ex:
```diff
@@ -37,12 +37,16 @@ defmodule Talon.Router do
-      get "/page/:page", TalonController, :page
-      get "/select_theme/:id", TalonController, :select_theme
+      get "/", TalonPageController, :dashboard
+      get "/dashboard", TalonPageController, :dashboard
```
I'm not sure we need a `/dashboard` route here. I think "/" and "/pages/:page/" should be sufficient.

smpallen99 commented on priv/templates/talon.gen.theme/admin_lte/templates/layout/sidebar.html.slim:
```diff
@@ -1,11 +1,18 @@
+      li.header.text-uppercase Pages
+      / TODO: use talon_context to get page_paths (DJS)
+      - resource_or_page = @talon_resource || @talon_page
+      = if resource_or_page do
```
why the check for resource_or_page here? Shouldn't listing pages and resources be independent of a specific page or resource? I can see the check for @resource for listing actions, but not for listing the collections.

smpallen99 commented on priv/templates/talon.gen.theme/admin_lte/views/dashboard_view.ex:
```diff
@@ -0,0 +1,4 @@
+defmodule <%= target_module %>.<%= web_namespace %>DashboardView do
```
Why do we need a dedicated dashboard view? should we not use a generic page_view in templates and create the view based on the page name?

smpallen99 commented on priv/templates/talon.new/talon/dashboard.ex:
```diff
@@ -0,0 +1,11 @@
+defmodule <%= base %>.Talon.Dashboard do
```
Same here. This should be a generic page.ex template that gets named by the generator.

smpallen99 commented on test/lib/dashboard_test.exs:
```diff
@@ -0,0 +1,22 @@
+  use TestTalon.ConnCase
+  require Logger
+
+  test "GET /talon/dashboard",  %{conn: conn} do
```
per earlier comment, I'm not sure why we want to  support this route.

dsnider committed:
Remove dashboard handler from PageController #40, #52

dsnider committed:
Remove dashboard routes from router #40

dsnider committed:
Remove dashboard cruft from Talon module #40

dsnider committed:
Generate dashboard and dashboard view from generic page templates #40, #52

dsnider committed:
Remove talon_page from conn and use only talon_resource #52

dsnider committed:
Only show Action section when relevant

dsnider committed:
Tighten sidebar partial using render_many

dsnider committed:
Alias the view module in sidebar template

dsnider committed:
Use unless empty check in sidebar

dsnider committed:
Add layout and sidebar helpers to LayoutView

dsnider committed:
Simplify sidebar link assign

dsnider committed:
Tighten nav resource link, add some comments

dsnider committed:
Merge with master

dsnider committed:
Merge with master

smpallen99 committed:
remove compiler warnings

smpallen99 committed:
Changed context to concern in page

- removes compliler warnings
- changed gettext domain Talon to talon
- added couple TODOs in code

smpallen99 committed:
fix broken tests from page concern option change

dsnider committed:
Change context to concern in several places

dsnider committed:
Merge branch 'feature/dashboard' of github.com:talonframework/talon into feature/dashboard
